aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -458,17 +525,22 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
   int64_t max_nnz = std::min(t_nnz, s_nnz);  // multiply by zero is zero, and can be dropped
   int64_t sparse_dim = src.sparse_dim();
   LongTensor t_indices = t._indices();
-  Tensor t_values = t._values();
   LongTensor src_indices = src._indices();
-  Tensor s_values = src._values();
   LongTensor r_indices = at::empty({sparse_dim, max_nnz}, t_indices.options());
-  Tensor r_values = new_values_with_size_of(t_values, max_nnz).zero_();
-  r.resize_as_(src);
-  get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
 
   int64_t match, d;
   int64_t r_i = 0, t_i = 0, s_i = 0;
 
+  auto commonDtype = promoteTypes(t_.scalar_type(), src_.scalar_type());
+  TORCH_CHECK(canCast(commonDtype, r.scalar_type()), "Can't convert result type ", commonDtype, " to output ", r.scalar_type());
+
+  Tensor t_values = t._values().to(commonDtype);
+  Tensor s_values = src._values().to(commonDtype);
```

Comment:
Let's say we promoted the tensors instead of promoting the `values()` of the tensors here. Then we wouldn't have had to change anything else in the implementation. What are the benefits of promoting the `values()` instead of the tensors?

Reply:
it's just simpler in aggregate, a small inplace change to existing code. existing code is:
```
Tensor t_values = t._values();
Tensor s_values = src._values();
```
This is:
```
Tensor t_values = t._values().to(commonDtype);
Tensor s_values = src._values().to(commonDtype);
```
The alternate change without trying to refactor anything else would be:
```
Tensor t_converted = t.to(commonDtype);
Tensor s_converted = src.to(commonDtype);
Tensor t_values = t_converted._values();
Tensor s_values = s_converted._values();
```
We still need to change the below because the op still needs to be done on `commonDtype` not the result type.  `r_values.scalar_type()` might be narrower than the common type, e.g. in `float_tensor *= double_tensor` the commonDtype is double but r_values.scalar_type() is float.